src/AppInstallerRepositoryCore/Microsoft/PredefinedInstalledSourceFactory.cpp

```diff
@@ -0,0 +1,175 @@
+
+                // Put installed packages into the index
+                if (filter == PredefinedInstalledSourceFactory::Filter::None || filter == PredefinedInstalledSourceFactory::Filter::ARP_System)
+                {
+                    PopulateIndexFromARP(index, HKEY_LOCAL_MACHINE);
+                }
+
+                if (filter == PredefinedInstalledSourceFactory::Filter::None || filter == PredefinedInstalledSourceFactory::Filter::ARP_User)
+                {
+                    PopulateIndexFromARP(index, HKEY_CURRENT_USER);
+                }
+
+                if (filter == PredefinedInstalledSourceFactory::Filter::None || filter == PredefinedInstalledSourceFactory::Filter::MSIX)
+                {
+                    PopulateIndexFromMSIX(index);
+                }
+
+                return std::make_shared<SQLiteIndexSource>(details, "*PredefinedInstalledSource", std::move(index), Synchronization::CrossProcessReaderWriteLock{}, true);
+            }
+
+            void Add(SourceDetails&, IProgressCallback&) override final
+            {
+                // Add should never be needed, as this is predefined.
+                THROW_HR(E_NOTIMPL);
+            }
+
+            void Update(const SourceDetails&, IProgressCallback&) override final
+            {
+                // Update could be used later, but not for now.
+                THROW_HR(E_NOTIMPL);
+            }
+
+            void Remove(const SourceDetails&, IProgressCallback&) override final
+            {
+                // Similar to add, remove should never be needed.
+                THROW_HR(E_NOTIMPL);
+            }
+        };
+    }
+
+    std::string_view PredefinedInstalledSourceFactory::FilterToString(Filter filter)
+    {
+        switch (filter)
+        {
+        case AppInstaller::Repository::Microsoft::PredefinedInstalledSourceFactory::Filter::None:
+            return "None"sv;
+        case AppInstaller::Repository::Microsoft::PredefinedInstalledSourceFactory::Filter::ARP_System:
+            return "ARP_System"sv;
+        case AppInstaller::Repository::Microsoft::PredefinedInstalledSourceFactory::Filter::ARP_User:
+            return "ARP_User"sv;
+        case AppInstaller::Repository::Microsoft::PredefinedInstalledSourceFactory::Filter::MSIX:
+            return "MSIX"sv;
+        default:
+            return "Unknown"sv;
+        }
+    }
+
+    PredefinedInstalledSourceFactory::Filter PredefinedInstalledSourceFactory::StringToFilter(std::string_view filter)
+    {
+        if (filter == FilterToString(Filter::ARP_System))
+        {
+            return Filter::ARP_System;
+        }
+        else if (filter == FilterToString(Filter::ARP_User))
+        {
+            return Filter::ARP_User;
+        }
+        else if (filter == FilterToString(Filter::MSIX))
+        {
+            return Filter::ARP_User;
+        }
+        else

```

Comment:
>else [](start = 8, length = 4)

Do we need unknown to be safe or we don't quite care?

Reply:
I don't think so.

---
In reply to: [507556942](https://github.com/microsoft/winget-cli/pull/598#discussion_r507556942) [](ancestors = 507556942)